Brokerages/Bitfinex/BitfinexWebsocketsBrokerage.Messaging.cs

```diff
@@ -0,0 +1,214 @@
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using QuantConnect.Data.Market;
+using QuantConnect.Logging;
+using QuantConnect.Orders;
+using System;
+using System.Collections.Generic;
+using System.Configuration;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using WebSocketSharp;
+
+namespace QuantConnect.Brokerages.Bitfinex
+{
+    public partial class BitfinexWebsocketsBrokerage
+    {
+
+        /// <summary>
+        /// Wss message handler
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        public void OnMessage(object sender, MessageEventArgs e)
+        {
+            try
+            {
+                var raw = JsonConvert.DeserializeObject<dynamic>(e.Data);
+
+                if (raw.Type == Newtonsoft.Json.Linq.JTokenType.Array)
+                {
+                    int id = raw[0];
+                    string term = raw[1];
+
+                    if (term == "hb")
+                    {
+                        //heartbeat
+                        _heartbeatCounter = DateTime.UtcNow;
+                        return;
+                    }
+                    else if (term == "tu" || term == "te")
+                    {
+                        //trade execution/update
+                        var data = raw[2].ToObject(typeof(string[]));
+                        PopulateTrade(data);
+                    }
+                    else if (term == "ws")
+                    {
+                        //wallet
+                        var data = raw[2].ToObject(typeof(string[][]));
+                        PopulateWallet(data);
+                    }
+                    else if (_channelId.ContainsKey(id) && _channelId[id] == "ticker")
+                    {
+                        //ticker
+                        PopulateTicker(e.Data);
+                        return;
+                    }
+                }
+                else if (raw.channel == "ticker" && raw.@event == "subscribed")
+                {
+                    if (!this._channelId.ContainsKey((int)raw.chanId))
+                    {
+                        this._channelId[(int)raw.chanId] = "ticker";
+                    }
+                }
+                else if (raw.chanId == 0)
+                {
+                    if (raw.status == "FAIL")
+                    {
+                        throw new Exception("Failed to authenticate with ws gateway");
+                    }
+                    Log.Trace("Successful wss auth");
```

Comment:
we like to prefix `Log.Trace` statements with the class and method, so this would become:
```C#
Log.Trace("BitfinexWebsocketsBrokerage.OnMessage(): Successful wss auth");
```
This makes it easy to know where log statements came from in the code